File: apps/frontend/src/services/settlements.ts
import { apiFetch } from "./api";

export interface Settlement {
  id: string;
  vendor_id: string;
  vendor_store_name?: string;
  period_start: string;
  period_end: string;
  total_redemptions: number;
  admin_fee: number;
  net_amount: number;
  status: "calculating" | "ready" | "processing" | "paid" | "cancelled";
  payout_date?: string;
  bank_transfer_reference?: string;
  created_at: string;
}

export interface SettlementDetail extends Settlement {
  vendor_bank_name?: string;
  vendor_bank_account?: string;
  vendor_account_holder?: string;
  admin_fee_percentage: number;
  breakdown: Array<{
    date: string;
    redemptions: number;
    amount: number;
  }>;
  notes?: string;
}

export interface SettlementListResponse {
  items: Settlement[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export async function getSettlements(
  page: number = 1,
  page_size: number = 20,
  status?: string
): Promise<SettlementListResponse> {
  const params = new URLSearchParams();
  params.append("page", page.toString());
  params.append("page_size", page_size.toString());
  if (status) {
    params.append("status", status);
  }
  return apiFetch(`/settlements/?${params.toString()}`);
}

export async function getSettlementDetail(settlementId: string): Promise<SettlementDetail> {
  return apiFetch(`/settlements/${settlementId}`);
}

export async function markSettlementPaid(
  settlementId: string,
  bankTransferReference: string,
  payoutDate?: string
): Promise<Settlement> {
  const data: Record<string, string> = {
    bank_transfer_reference: bankTransferReference,
  };
  if (payoutDate) {
    data.payout_date = payoutDate;
  }
  return apiFetch(`/settlements/${settlementId}/mark-paid`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function requestSettlementPayout(settlementId: string): Promise<Settlement> {
  return apiFetch(`/settlements/${settlementId}/request-payout`, {
    method: "POST",
  });
}

export async function calculateSettlements(
  periodStart: string,
  periodEnd: string,
  vendorId?: string
): Promise<{ settlements_created: number; total_amount: number }> {
  const data: Record<string, string> = {
    period_start: periodStart,
    period_end: periodEnd,
  };
  if (vendorId) {
    data.vendor_id = vendorId;
  }
  return apiFetch("/settlements/calculate", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * Export settlements as CSV
 * @param format - 'csv' (pdf not yet supported)
 * @param startDate - Optional start date filter
 * @param endDate - Optional end date filter
 * @returns Blob of the CSV file
 */
export async function exportSettlements(
  format: "csv" | "pdf" = "csv",
  startDate?: string,
  endDate?: string
): Promise<Blob> {
  const params = new URLSearchParams();
  params.append("format", format);
  if (startDate) params.append("start_date", startDate);
  if (endDate) params.append("end_date", endDate);

  const response = await fetch(
    `${import.meta.env.VITE_API_BASE_URL}/settlements/export?${params.toString()}`,
    {
      method: "GET",
      headers: {
        Authorization: `Bearer ${localStorage.getItem("access_token")}`,
      },
    }
  );

  if (!response.ok) {
    throw new Error("Gagal mengekspor settlement");
  }

  return response.blob();
}

File: apps/backend/app/api/settlements.py
"""
Settlement Router
Handles settlement listing, detail, and calculation
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
from datetime import date, datetime

from app.database import get_db
from app.services.settlement_service import SettlementService
from app.middleware.auth import get_current_user, AuthenticatedUser, RequireRole
from app.schemas.settlement import (
    SettlementResponse,
    SettlementDetailResponse,
    SettlementListResponse,
    SettlementCalculateRequest,
    SettlementCalculateResponse,
    SettlementQueryParams,
    SettlementMarkPaidRequest,
    SettlementMarkPaidResponse,
)
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/settlements", tags=["settlements"])


@router.get("/", response_model=SettlementListResponse)
async def list_settlements(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status_filter: Optional[str] = Query(None, alias="status"),
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """List settlements for vendor (own) or admin (all)"""
    is_admin = current_user.role == "admin"
    params = SettlementQueryParams(
        page=page,
        page_size=page_size,
        status=status_filter,
        start_date=start_date,
        end_date=end_date,
    )

    settlements = SettlementService.get_settlements(
        db, current_user.user_id, is_admin, params
    )
    total = SettlementService.get_settlements_count(
        db, current_user.user_id, is_admin, params
    )
    total_pages = (total + page_size - 1) // page_size if total > 0 else 0

    items = []
    for s in settlements:
        s_dict = SettlementResponse.model_validate(s).model_dump()
        if s.vendor_profile:
            s_dict["vendor_store_name"] = s.vendor_profile.store_name
        items.append(SettlementResponse(**s_dict))

    return SettlementListResponse(
        items=items, total=total, page=page, page_size=page_size, total_pages=total_pages
    )


@router.get("/{settlement_id}", response_model=SettlementDetailResponse)
async def get_settlement(
    settlement_id: str,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """Get settlement detail with breakdown"""
    is_admin = current_user.role == "admin"
    settlement = SettlementService.get_settlement_by_id(
        db, settlement_id, current_user.user_id, is_admin
    )

    if not settlement:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Settlement not found",
        )

    s_dict = SettlementDetailResponse.model_validate(settlement).model_dump()
    if settlement.vendor_profile:
        s_dict["vendor_store_name"] = settlement.vendor_profile.store_name
        s_dict["vendor_bank_name"] = settlement.vendor_profile.bank_name
        s_dict["vendor_bank_account"] = settlement.vendor_profile.bank_account_number
        s_dict["vendor_account_holder"] = settlement.vendor_profile.bank_account_holder

    s_dict["breakdown"] = SettlementService.get_settlement_breakdown(db, settlement)
    s_dict["admin_fee_percentage"] = 1.0

    return SettlementDetailResponse(**s_dict)


@router.post("/calculate", response_model=SettlementCalculateResponse)
async def calculate_settlements(
    data: SettlementCalculateRequest,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(RequireRole(["admin"])),
):
    """Trigger settlement calculation (admin only)"""
    result = SettlementService.calculate_settlements(
        db,
        period_start=data.period_start,
        period_end=data.period_end,
        vendor_id=data.vendor_id,
    )
    return SettlementCalculateResponse(**result)


@router.post("/{settlement_id}/request-payout", response_model=SettlementMarkPaidResponse)
async def request_settlement_payout(
    settlement_id: str,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """Request payout for a settlement (vendor requests payout)"""
    is_admin = current_user.role == "admin"
    settlement = SettlementService.get_settlement_by_id(
        db, settlement_id, current_user.user_id, is_admin
    )

    if not settlement:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Settlement not found",
        )

    if settlement.status not in ["ready"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Cannot request payout for settlement with status: {settlement.status}",
        )

    settlement.status = "processing"
    settlement.updated_at = datetime.now()
    db.add(settlement)
    db.commit()
    db.refresh(settlement)

    logger.info(f"Settlement {settlement_id} payout requested by vendor")

    s_dict = SettlementMarkPaidResponse.model_validate(settlement).model_dump()
    if settlement.vendor_profile:
        s_dict["vendor_store_name"] = settlement.vendor_profile.store_name

    return SettlementMarkPaidResponse(**s_dict)


@router.post("/{settlement_id}/mark-paid", response_model=SettlementMarkPaidResponse)
async def mark_settlement_paid(
    settlement_id: str,
    data: SettlementMarkPaidRequest,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(RequireRole(["admin"])),
):
    """Mark settlement as paid (admin only)"""
    is_admin = current_user.role == "admin"
    settlement = SettlementService.get_settlement_by_id(
        db, settlement_id, current_user.user_id, is_admin
    )

    if not settlement:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Settlement not found",
        )

    if settlement.status != "ready":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Settlement must be in 'ready' status, current status: {settlement.status}",
        )

    # Update settlement status
    settlement.status = "paid"
    settlement.payout_date = data.payout_date or date.today()
    settlement.bank_transfer_reference = data.bank_transfer_reference
    settlement.updated_at = datetime.now()

    db.add(settlement)
    db.commit()
    db.refresh(settlement)

    logger.info(
        f"Settlement {settlement_id} marked as paid with reference: {data.bank_transfer_reference}"
    )

    s_dict = SettlementMarkPaidResponse.model_validate(settlement).model_dump()
    if settlement.vendor_profile:
        s_dict["vendor_store_name"] = settlement.vendor_profile.store_name

    return SettlementMarkPaidResponse(**s_dict)


@router.get("/export")
async def export_settlements(
    format: str = Query("csv", regex="^(csv|pdf)$"),
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(get_current_user),
):
    """Export settlements as CSV or PDF"""
    from fastapi.responses import StreamingResponse
    import csv
    import io
    
    # Get settlements for current user
    is_admin = current_user.role == "admin"
    params = SettlementQueryParams(
        page=1,
        page_size=1000,  # Get all for export
        start_date=start_date,
        end_date=end_date,
    )
    
    settlements = SettlementService.get_settlements(
        db, current_user.user_id, is_admin, params
    )
    
    if format.lower() == "csv":
        # Generate CSV
        output = io.StringIO()
        writer = csv.writer(output)
        
        # Header
        writer.writerow([
            "ID", "Period Start", "Period End", "Total Orders", "Total Amount",
            "Admin Fee", "Net Amount", "Status", "Created At"
        ])
        
        # Data
        for s in settlements:
            writer.writerow([
                str(s.id),
                s.period_start.isoformat() if s.period_start else "",
                s.period_end.isoformat() if s.period_end else "",
                s.total_orders,
                str(s.total_amount),
                str(s.admin_fee),
                str(s.net_amount),
                s.status,
                s.created_at.isoformat() if s.created_at else "",
            ])
        
        output.seek(0)
        
        return StreamingResponse(
            io.BytesIO(output.getvalue().encode()),
            media_type="text/csv",
            headers={"Content-Disposition": "attachment; filename=settlements.csv"}
        )
    
    else:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="PDF format not yet implemented. Use 'csv'."
        )
